### PR TITLE
Committed Sync

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: radixdlt/src/main/java/com/radixdlt/middleware2/network/GetVerticesResponseMessage.java [hunk 1]
```diff
@@ -31,23 +33,33 @@
...
 		this.vertices = Objects.requireNonNull(vertices);
 	}
 
+	public Hash getVertexId() {
+		return vertexId;
+	}
+
 	public List<Vertex> getVertices() {
-		return vertices;
+		return vertices == null ? Collections.emptyList() : vertices;
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/middleware2/network/GetVerticesResponseMessage.java]: Would prefer this to be returning an immutable list, so that this object remained immutable.

**Review by clstrfsck — DISMISSED** [review]

**Review by clstrfsck — CHANGES_REQUESTED** [review]
Looks OK, but tests are failing for some reason.

File: radixdlt/src/integration/java/com/radixdlt/consensus/simulation/BFTSimulatedTest.java [hunk 2]
```diff
@@ -167,7 +168,7 @@ public static Builder builder() {
 	 * @param timeUnit time unit of duration
 	 * @return map of check results
 	 */
-	public Map<String, Boolean> run(long duration, TimeUnit timeUnit) {
+	public Map<String, Optional<BFTCheckError>> run(long duration, TimeUnit timeUnit) {
```
Comment (flotothemoon) [inline radixdlt/src/integration/java/com/radixdlt/consensus/simulation/BFTSimulatedTest.java:171]: Yay!

File: radixdlt/src/integration/java/com/radixdlt/consensus/simulation/checks/SafetyCheck.java [hunk 3]
```diff
@@ -17,34 +17,105 @@
...
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Collectors;
 
 /**
- * Checks that nodes are committed on the same vertex.
+ * Checks that nodes do not commit on conflicting vertices
  */
 public class SafetyCheck implements BFTCheck {
 
+	private static Observable<BFTCheckError> differentVerticesError(Vertex vertex, Vertex currentVertex) {
```
Comment (flotothemoon) [inline radixdlt/src/integration/java/com/radixdlt/consensus/simulation/checks/SafetyCheck.java]: "conflicting" might be more clear

File: radixdlt/src/integration/java/com/radixdlt/consensus/simulation/checks/SafetyCheck.java [hunk 4]
```diff
@@ -17,34 +17,105 @@
...
+					highest.set(vertex.getView());
 				}
+
+				// Clean up old vertices so that we avoid consuming too much memory
+				lastCommittedByNode.put(node, vertex);
+				final View lowest = lastCommittedByNode.values().stream()
+					.map(Vertex::getView)
+					.reduce((v0, v1) -> v0.compareTo(v1) < 0 ? v0 : v1)
+					.orElse(View.genesis());
+				final Set<View> viewsToRemove = committedVertices.keySet().stream()
```
Comment (Sajjon) [inline radixdlt/src/integration/java/com/radixdlt/consensus/simulation/checks/SafetyCheck.java:114]: Skip using sets and use array instead to get deterministic behaviour? Or is it important to have non-determinism here? :) For uniqueness you can use `.distinct()` I think?

File: radixdlt/src/integration/java/com/radixdlt/consensus/simulation/checks/SafetyCheck.java [hunk 5]
```diff
@@ -17,34 +17,105 @@
...
+				node -> network.getVertexStore(node)
+					.lastCommittedVertex().map(v -> Pair.of(node, v)))
+				.collect(Collectors.toList())
 		)
-			.map(committedVertices -> committedVertices.distinct().collect(Collectors.toList()))
-			.concatMap(committedVertices -> {
-				if (committedVertices.size() != 1) {
-					return Observable.just(new BFTCheckError("Committed vertices don't match: " + committedVertices));
+			.flatMap(nodeAndVertex -> {
+				final ECKeyPair node = nodeAndVertex.getFirst();
```
Comment (Sajjon) [inline radixdlt/src/integration/java/com/radixdlt/consensus/simulation/checks/SafetyCheck.java:86]: Pretty long anonymous function inside this `flatMap`, making it hard to follow the reactive code, maybe break out to helper method? :)

File: radixdlt/src/main/java/com/radixdlt/consensus/SyncedStateComputer.java [hunk 6]
```diff
@@ -47,9 +46,11 @@
 	 *
 	 * @param targetStateVersion the target state version
 	 * @param target list of targets as hint of which peer has the state
+	 * @param opaque some opaque client object which will be passed in a sync
+	 * message if this returns false
 	 * @return a completable which completes when sync is complete
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/SyncedStateComputer.java]: update documentation, no longer a completable

**Review by Sajjon — COMMENTED** [review]
When Rx anonymous functions become too long it becomes hard to follow the reactive flow, when possible, I think it would be advantageous to break out helper method, what do you think?

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java [hunk 7]
```diff
@@ -115,73 +182,164 @@ public void processGetVerticesRequest(GetVerticesRequest request) {
...
 
-	private void doSync(QuorumCertificate qc, ECPublicKey author) {
-		final Hash vertexId = qc.getProposed().getId();
+	public void processGetVerticesResponse(GetVerticesResponse response) {
+		log.info("SYNC_VERTICES: Received GetVerticesResponse {}", response);
+
+		final Hash syncTo = (Hash) response.getOpaque();
+		SyncState syncState = syncing.get(syncTo);
+		if (syncState == null) {
+			return;
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java]: Does this signal an invalid or malicious response? Can we do something here besides ignoring it (e.g. log it)?

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java [hunk 8]
```diff
@@ -115,73 +182,164 @@ public void processGetVerticesRequest(GetVerticesRequest request) {
...
 		if (addQC(qc)) {
 			return true;
 		}
 
-		if (author != null) {
-			this.doSync(qc, author);
+		log.info("SYNC_TO_QC: Need sync: {} {}", qc, committedQC);
+
+		if (author == null) {
+			throw new IllegalStateException("Syncing required but author doesn't exist");
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java]: "author wasn't provided"? "doesn't exist" sounds like author is non-null but we couldn't find it in some way to me

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java [hunk 9]
```diff
@@ -69,40 +71,105 @@
...
 		}
+	}
 
-		this.rootId.set(genesisVertex.getId());
-		this.vertices.put(genesisVertex.getId(), genesisVertex);
-		this.lastCommittedVertex.onNext(genesisVertex);
+	private enum SyncStage {
+		PREPARING,
+		GET_COMMITTED_VERTICES,
+		SYNC_TO_COMMIT, GET_QC_VERTICES
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java]: should be on a new line according to our style guide
```suggestion
		SYNC_TO_COMMIT,
		GET_QC_VERTICES
```

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java [hunk 10]
```diff
@@ -115,73 +182,164 @@ public void processGetVerticesRequest(GetVerticesRequest request) {
...
+		}
+
+		syncState.setSyncStage(SyncStage.GET_COMMITTED_VERTICES);
+		syncing.put(vertexId, syncState);
+
+		log.info("SYNC_VERTICES: Committed: Sending initial GetVerticesRequest {}", syncState.getCommittedQC());
+		syncVerticesRPCSender.sendGetVerticesRequest(vertexId, syncState.author, 3, vertexId);
+	}
+
+
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java]: extra newline here

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java [hunk 11]
```diff
@@ -194,18 +352,23 @@ private boolean addQC(QuorumCertificate qc) {
...
-				|| highestCommitted.getCommitted().get().getStateVersion() < vertexMetadata.getStateVersion()) {
+			QuorumCertificate highestCommitted = highestCommittedQC.get();
+			Optional<VertexMetadata> highest = highestCommitted.getCommitted();
+			if (highest.isPresent()) {
+				if (highest.get().getView().compareTo(vertexMetadata.getView()) < 0) {
+					this.highestCommittedQC.set(qc);
+				}
+			} else {
 				this.highestCommittedQC.set(qc);
 			}
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java]: Could be simplified:
```suggestion
			if (!highest.isPresent() || highest.get().getView().compareTo(vertexMetadata.getView()) < 0) {
				this.highestCommittedQC.set(qc);
			}
```

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java [hunk 12]
```diff
@@ -69,41 +71,116 @@
...
-			syncedStateComputer.execute(committedGenesis);
+		for (Vertex vertex : vertices) {
+			if (!addQC(vertex.getQC())) {
+				throw new IllegalStateException(String.format("Missing qc=%s", vertex.getQC()));
+			}
+
+			try {
+				insertVertexInternal(vertex);
+			} catch (VertexInsertionException e) {
+				throw new IllegalStateException("Could not insert vertex " + vertex);
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java]: Adding the insertion error as a cause to the exception would be nice

File: radixdlt/src/integration/java/com/radixdlt/consensus/simulation/checks/SafetyCheck.java (hunk 4 above)
Comment (talekhinezh) [reply radixdlt/src/integration/java/com/radixdlt/consensus/simulation/checks/SafetyCheck.java:114]: Not sure what that gains since deterministic iteration order doesn't matter in this logic.
Also, there's already a lot of "non-determinism" in here like iterating through keys of the map or `.reduce` function above.

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java [hunk 13]
```diff
@@ -115,71 +192,155 @@ public void processGetVerticesRequest(GetVerticesRequest request) {
...
+		syncState.setSyncStage(SyncStage.GET_QC_VERTICES);
+		log.info("SYNC_VERTICES: QC: Sending initial GetVerticesRequest for qc={}", syncState.getQC());
 		syncVerticesRPCSender.sendGetVerticesRequest(vertexId, syncState.author, 1, vertexId);
 	}
 
+	private void doCommittedSync(SyncState syncState) {
+		final Hash vertexId = syncState.getQC().getProposed().getId();
+		syncState.setSyncStage(SyncStage.GET_COMMITTED_VERTICES);
+		log.info("SYNC_VERTICES: Committed: Sending initial GetVerticesRequest {}", syncState.getCommittedQC());
+		syncVerticesRPCSender.sendGetVerticesRequest(vertexId, syncState.author, 3, vertexId);
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java]: What is the reason for the magic 3 here?

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java [hunk 14]
```diff
@@ -115,71 +192,155 @@ public void processGetVerticesRequest(GetVerticesRequest request) {
...
-			// failed
-			return;
+		long stateVersion = syncState.committedVertexMetadata.getStateVersion();
+		List<ECPublicKey> signers = Collections.singletonList(syncState.author);
+		syncState.fetched.addAll(response.getVertices());
+
+		if (syncedStateComputer.syncTo(stateVersion, signers, syncTo)) {
+			rebuildAndSyncQC(syncState);
+		} else {
+			syncState.syncStage = SyncStage.SYNC_TO_COMMIT;
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java]: Should this be using the setter for `syncStage`?

**Review by clstrfsck — APPROVED** [review]

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java (hunk 13 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java]: want the three children of committedQC so that we can construct a valid committed root

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java (hunk 14 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java]: yes

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java (hunk 11 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java]: cool

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java (hunk 7 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java]: no shouldn't be a malicious response since the RPC layer should be handling matching requests/responses. This signals that the sync was handled by a different sync and was already processed.

File: radixdlt/src/integration/java/com/radixdlt/consensus/simulation/checks/SafetyCheck.java (hunk 5 above)
Comment (talekhinezh) [reply radixdlt/src/integration/java/com/radixdlt/consensus/simulation/checks/SafetyCheck.java:86]: Breaking it out I think will make it harder to follow since this check requires a bit of state (e.g. committedVertices map). Could break it out if there was a bigger refactor to enable moving some state to the class level but that would need to be in the future.

**Review by flotothemoon — APPROVED** [review]
